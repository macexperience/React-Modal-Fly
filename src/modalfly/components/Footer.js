import { useState, useEffect } from 'react';
import ReactDOM from 'react-dom';
import './styles.css';

export function Footer(props) {
    const [footerRef, setFooterRef] = useState(null);

    //When component  mounts, set footer reference
    useEffect(() => {
        //Query all footer areas    
        const elements = document.getElementsByClassName('mf-footer-area');
        //Select the most recent addition, the div with no children
        let footerRef;
        for (let i = 0; i < elements.length; i++) {
            if (!elements[i].hasChildNodes()) footerRef = elements[i];
        }

        var zIndex = Math.max.apply(null, Array.prototype.map.call(document.querySelectorAll('*'), function(el) {
            return +el.style.zIndex;
          })) + 10;
        footerRef.style.zIndex = zIndex * 2;

        setFooterRef(footerRef);
    }, []);

    if (footerRef) {
        //Render the buttons in the mf-footer div ('footerRef')
        return ReactDOM.createPortal(props.children, footerRef);
    } else {
        return null;
    }
}