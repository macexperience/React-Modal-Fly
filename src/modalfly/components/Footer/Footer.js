import { useState, useEffect } from 'react';
import ReactDOM from 'react-dom';

export function Footer(props) {
    const [footerRef, setFooterRef] = useState(null);

    //When component  mounts, set footer reference
    useEffect(() => {
        //Query all footer areas    
        const elements = document.getElementsByClassName('mf-footer-area');
        //No do not set footer reference if elements is empty
        if (elements.length === 0) {
            return;
        }
        //Select the most recent addition, the div with no children
        let footerRef;
        for (let i = 0; i < elements.length; i++) {
            if (!elements[i].hasChildNodes()) footerRef = elements[i];
        }

        var zIndex = Math.max.apply(null, Array.prototype.map.call(document.querySelectorAll('*'), function(el) {
            return +el.style.zIndex;
          })) + 10;
        footerRef.style.zIndex = zIndex * 2;

        setFooterRef(footerRef);
    }, []);

    if (footerRef) {
        //Render the buttons in the mf-footer div ('footerRef')
        return ReactDOM.createPortal(props.children, footerRef);
    } else {
        return null;
    }
}