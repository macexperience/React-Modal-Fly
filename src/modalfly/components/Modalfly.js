import React, { useReducer, useEffect, useState } from 'react';
import ReactDOM from 'react-dom';
import './styles.css';
import { reducer, initialState } from '../hooks/reducer';
import * as actions from '../hooks/actions';
import StepContext from '../hooks/StepContext';
import { CSSTransition } from 'react-transition-group';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import { isFunction, getModalStyles } from '../helpers/helpers'
import * as styles from '../styles/styles';

export function Modalfly(props) {
    //Create reducer
    const [store, dispatch] = useReducer(reducer, initialState);

    const [containerRef, setContainerRef] = useState(null);

    //Get container reference
    useEffect(() => {
        let containerRef = document.getElementById('modalfly-container');
        if (containerRef) {
            setContainerRef(containerRef);
        } else {
            setContainerRef(false);
        }
    });


    //Set root style
    let wrapperClass = classNames({
        'mf-wrapper': true,
        'mf-small': props.size === 'small',
        'mf-meduim': props.size === 'medium',
        'mf-large': props.size === 'large',
        'mf-xl': props.size === 'extraLarge'
    });

    const mfStyle = getModalStyles(props);
    console.log('mfStyle:', mfStyle);


    //Set Footer style
    let footerClass = classNames({
        'mf-footer-area': true,
        'mf-footer-left': props.footerContent === 'left',
        'mf-footer-center': props.footerContent === 'center',
        'mf-footer-right': props.footerContent === 'right',
        'mf-footer-space-between': props.footerContent === 'spaceBetween',
        'mf-footer-space-around': props.footerContent === 'spaceAround',
        'mf-footer-space-evenly': props.footerContent === 'spaceEvenly'
    });

    /* Do not show component if show is false */
    if (!props.show) {
        return null;
    }
    //Check mode
    const inWorkflowMode = props.workflow && props.children;

    //If in workflow mode, count steps and get step titles
    let titlesList = []
    let currentStepCount;
    if (inWorkflowMode) {
        //Count the current steps
        currentStepCount = props.children.props.children.length;
        if (currentStepCount === 0) return new Error(`Modalfly workflow mode requires at least one Step.`);
        //Get titles
        const divSteps = props.children.props.children;
        titlesList = divSteps.map(step => {
            return step.props.title && typeof step.props.title === 'string' ? step.props.title : 'Attention'
        });
    }

    //Check for onClose param
    const onCloseIsValid = isFunction(props.onClose);
    const displayCloseIcon = props.onClose && onCloseIsValid;

    const closeIcon = () => {
        if (displayCloseIcon) {
            return (<span onClick={onCancel} className="mf-close-icon">&times;</span>);
        } else {
            return null;
        }
    }

    // Generate progress circles for workflow mode
    const progressCircles = () => {
        //Return if not in workflow mode
        if (!inWorkflowMode || props.hideProgress) return null;

        let circles = [];
        for (let i = 0; i < currentStepCount; i++) {
            if (i === store.currentStep) {
                circles.push(<span key={i} className='mf-dots active'></span>);
            } else {
                circles.push(<span key={i} className='mf-dots'></span>);
            }
        }
        return circles;
    }

    const titlesGenerator = () => {
        if (props.title && typeof props.title === 'string') {
            return props.title;
        }
        else if (Array.isArray(titlesList) && titlesList.length !== 0) {
            return titlesList[store.currentStep];
        } else {
            return 'Attention';
        }
    }

    const onCancel = () => {
        dispatch({ type: actions.RESET });
        //Close modal
        props.onClose();
    }

    const contentArea = () => {
        return (
            <div className="mf-content-area" currentstep={store.currentStep}>
                {props.children}
            </div>
        );
    }

    const modal = (
        <StepContext.Provider value={{ store, dispatch }}>
            <CSSTransition
                unmountOnExit
                in={props.show}
                timeout={{ appear: 0, enter: 0, exit: 500 }}
                classNames="mf"
                appear
            >
                <div id='mf-wrapper' className={wrapperClass}>

                    <div className="mf-header-area">
                        <h3>
                            {titlesGenerator()}
                        </h3>
                    </div>

                    {closeIcon()}

                    {contentArea()}

                    <div className="mf-progress-area">
                        {progressCircles()}
                    </div>

                    <div id='mf-footer' className={footerClass}></div>
                </div>
            </CSSTransition>
        </StepContext.Provider>
    );

    if (containerRef) {
        return ReactDOM.createPortal(modal, containerRef);
    }

    return modal;
}

Modalfly.propTypes = {
    //Validate children 
    children: (props, propName, componentName) => {
        let error;
        const prop = props[propName];
        // console.log('React.Children.count:', React.Children.toArray())
        React.Children.forEach(prop, (child) => {
            //If in 'workflow' mode, only allow child of Steps
            if (props.workflow && child.type.name !== 'Steps') {
                error = new Error(`${componentName} only accepts children of type 'Option' in workflow mode.`);
            }
        });
        return error;
    },
    //Modal sizes
    size: PropTypes.oneOf(['small', 'medium', 'large', 'extraLarge']),
    //Determines how the footer contents will be justified
    footerContent: PropTypes.oneOf(['left', 'center', 'right', 'spaceBetween', 'spaceAround', 'spaceEvenly']),
    //Show/hide progress circles
    hideProgress: PropTypes.bool,
    //Sets modal to workflow mode
    workflow: PropTypes.bool,
    onClose: (props, propName, componentName) => {
        if (props.onClose && !isFunction(props.onClose)) {
            return new Error(`${componentName} onClose prop must be a function.`)
        }
    }
};