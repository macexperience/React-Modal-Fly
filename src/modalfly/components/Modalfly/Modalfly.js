import React, { useReducer, useEffect, useState } from 'react';
import ReactDOM from 'react-dom';
import { reducer, initialState } from '../../hooks/reducer';
import * as actions from '../../hooks/actions';
import StepContext from '../../hooks/StepContext';
import { Transition } from 'react-transition-group';
import PropTypes from 'prop-types';
import { isFunction, getContainerStyle, getHeaderStyle, getFooterStyle } from '../../helpers/helpers'
import * as styles from '../../styles/styles';

export function Modalfly(props) {
    //Create reducer
    const [store, dispatch] = useReducer(reducer, initialState);

    const [containerRef, setContainerRef] = useState(null);

    //Get container reference
    useEffect(() => {
        let containerRef = document.getElementById('modalfly-container');
        if (containerRef) {
            setContainerRef(containerRef);
        } else {
            setContainerRef(false);
        }
    }, [props.show]);

    //Get container and footer styles
    const mfStyle = getContainerStyle(props);
    const headerStyle = getHeaderStyle(props);
    const footerStyle = getFooterStyle(props);

    if (!props.show) {
        return null;
    }

    //Check mode
    const inWorkflowMode = props.workflow && props.children;

    //If in workflow mode, count steps and get step titles
    let titlesList = []
    let currentStepCount;
    if (inWorkflowMode) {
        const children = props.children.props.children
        //Ensure there exist at least 2 steps for workflow mode
        const isAnObject = typeof children === 'object' && children !== null;
        const isAnArray = Array.isArray(children);
        if (isAnObject && !isAnArray) {
            return (<h3 style={{ color: "red" }}>'Modalfly' Error: At least two 'Step' components are required for workflow mode.</h3>)
        }
        //Count the current steps
        currentStepCount = children.length;
        if (currentStepCount === 0) return new Error(`Modalfly workflow mode requires at least one Step.`);
        //Get titles
        titlesList = children.map(step => {
            return step.props.title && typeof step.props.title === 'string' ? step.props.title : 'Attention'
        });
    }

    //Check for onClose param
    const onCloseIsValid = isFunction(props.onClose);
    const displayCloseIcon = props.onClose && onCloseIsValid;

    const closeIcon = () => {
        if (displayCloseIcon) {
            if (props.closeBtnClassName) {
                return (<i onClick={onCancel} className={props.closeBtnClassName ? props.closeBtnClassName : ''}></i>);
            }
            return (<i onClick={onCancel} style={styles.closeIcon}>&times;</i>);
        } else {
            return null;
        }
    }

    // Generate progress circles for workflow mode
    const progressCircles = () => {
        //Return if not in workflow mode
        if (!inWorkflowMode || props.hideProgress) return null;

        let circles = [];
        for (let i = 0; i < currentStepCount; i++) {
            if (i === store.currentStep) {
                circles.push(<span key={i} className='mf-dots active'></span>);
            } else {
                circles.push(<span key={i} className='mf-dots'></span>);
            }
        }
        return circles;
    }

    const titlesGenerator = () => {
        if (props.title && typeof props.title === 'string') {
            return props.title;
        }
        else if (Array.isArray(titlesList) && titlesList.length !== 0) {
            return titlesList[store.currentStep];
        } else {
            return 'Attention';
        }
    }

    const onCancel = () => {
        //Check 'resetSteps' prop, if 'false', do not reset steps.
        let reset = props.resetSteps === undefined ? true : props.resetSteps;
        //Always reset when in 'single view' mode
        if (props.workflow === undefined) {
            reset = true;
        }
        //Call client's onClose handler
        props.onClose();
        //Reset workflow steps
        if (reset) {
            dispatch({ type: actions.RESET });
        }
    }

    const contentArea = () => {
        return (
            <div style={styles.contentArea} currentstep={store.currentStep}>
                {props.children}
            </div>
        );
    }

    const modal = (
        <>
            <StepContext.Provider value={{ store, dispatch }}>
                <Transition
                    unmountOnExit
                    in={props.show}
                    timeout={{ appear: 0, enter: 600, exit: 500 }}
                    appear
                >
                    {
                        state => (
                            <div id='mf-wrapper' style={{ ...mfStyle, ...styles.mfTransitionStyles[state] }}>
                                {/* Header div */}
                                <div
                                    className={props.headerClassName ? props.headerClassName : ''}
                                    style={headerStyle}>
                                    <h3 style={styles.headerH3}>
                                        {titlesGenerator()}
                                    </h3>
                                    {closeIcon()}
                                </div>
                                {/* Main Content div */}
                                {contentArea()}

                                {/* Progress circles div */}
                                <div style={styles.progressArea}>
                                    {progressCircles()}
                                </div>
                                {/* Footer div */}
                                <div
                                    id='mf-footer'
                                    style={footerStyle}
                                    className={props.footerClassName ? `mf-footer-area ${props.footerClassName}` : 'mf-footer-area'}></div>
                            </div>
                        )
                    }

                </Transition>
            </StepContext.Provider>
        </>
    );

    if (containerRef) {
        return ReactDOM.createPortal(modal, containerRef);
    }

    return modal;
}

Modalfly.propTypes = {
    //Validate children 
    children: (props, propName, componentName) => {
        let error;
        const prop = props[propName];
        React.Children.forEach(prop, (child) => {
            //If in 'workflow' mode, only allow child of Steps
            if (props.workflow && child.type.name !== 'Steps') {
                error = new Error(`${componentName} only accepts children of type 'Option' in workflow mode.`);
            }
        });
        return error;
    },
    //Determines how the footer contents will be justified
    footerContent: PropTypes.oneOf(['left', 'center', 'right', 'spaceBetween', 'spaceAround', 'spaceEvenly']),
    //Show/hide progress circles
    hideProgress: PropTypes.bool,
    //Modal sizes
    size: PropTypes.oneOf(['small', 'medium', 'large', 'extraLarge']),
    //ModalContainer style
    style: PropTypes.object,
    //Footer style
    footerStyle: PropTypes.object,
    //Header class name
    headerClassName: PropTypes.string,
    //Footer class name
    footerClassName: PropTypes.string,
    //Close button class name
    closeBtnClassName: PropTypes.string,
    //Sets modal to workflow mode
    workflow: PropTypes.bool,
    onClose: (props, propName, componentName) => {
        if (props.onClose && !isFunction(props.onClose)) {
            return new Error(`${componentName} onClose prop must be a function.`)
        }
    }
};