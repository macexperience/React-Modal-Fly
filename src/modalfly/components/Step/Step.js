import React from 'react';

export function Step(props) {
    return (
        <div title={props.title} className={props.className} style={props.style}>
            {props.children}
        </div>
    );
};