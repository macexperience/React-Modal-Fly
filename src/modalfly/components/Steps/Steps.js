import React, { useContext, useState } from 'react';
import StepContext from '../../hooks/StepContext';
import * as types from '../../hooks/actions';

export function Steps(props) {
    const stepContext = useContext(StepContext);
    const [stepData, setStepData] = useState({});

    const totalSections = props.children ? props.children.length : 0;
    //Do not render if there are no children
    if (totalSections === 0) {
        return null;
    }
    //Expose next function
    Steps.next = (e) => {
        if (stepContext.store.currentStep < totalSections - 1) {
            stepContext.dispatch({ type: types.STEP_FORWARD, data: stepData });
        } else if (stepContext.store.currentStep === totalSections - 1) {
            stepContext.dispatch({ type: types.PERSIST_STEP_DATA, data: stepData });
        }
        //Clear state
        setStepData({});
    }

    Steps.previous = e => {
        if (stepContext.store.currentStep !== 0) {
            stepContext.dispatch({ type: types.STEP_BACKWARD });
        }
    }

    Steps.complete = e => {
        stepContext.dispatch({ type: types.RESET })
    }
    
    return props.children[stepContext.store.currentStep];
};