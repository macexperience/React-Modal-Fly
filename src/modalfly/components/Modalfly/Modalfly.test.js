import React from 'react';
import renderer from 'react-test-renderer';
import { Steps } from '../Steps/Steps';
import { Step } from '../Step/Step';
import { Footer } from '../Footer/Footer';
import { Modalfly } from './Modalfly';

describe('<Modalfly />', () => {
    it('renders an empty modal', () => {
        expect.assertions(1);
        const component = renderer.create(<Modalfly show={true} />);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders an single view modal', () => {
        expect.assertions(1);
        const component = renderer.create(
            <Modalfly show={true}>
                <div>
                    <h3>Would you like to overwrite?</h3>
                    <p>Lorem ipsum dolor sit amet consectetur adipisicing elit.</p>
                </div>
                <Footer>
                    <button>Close</button>
                    <button>Thanks</button>
                </Footer>
            </Modalfly>
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});