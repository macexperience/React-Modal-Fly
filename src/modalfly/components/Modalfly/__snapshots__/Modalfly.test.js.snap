// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modalfly /> renders an empty modal 1`] = `
<div
  id="mf-wrapper"
  style={
    Object {
      "background": "white",
      "borderRadius": "5px",
      "boxShadow": "0 5px 8px 0 rgba(0, 0, 0, 0.2), 0 7px 20px 0 rgba(0, 0, 0, 0.11)",
      "display": "grid",
      "fontSize": "1em",
      "gridTemplateColumns": "2fr 14fr 2fr",
      "gridTemplateRows": "70px auto 10px 80px",
      "left": "50%",
      "maxHeight": "100%",
      "maxWidth": "100%",
      "minHeight": "35%",
      "opacity": 0,
      "overflow": "visible",
      "position": "fixed",
      "top": "40%",
      "transform": "translate(-50%, -50%)",
      "transition": "opacity 500ms",
      "width": "500px",
    }
  }
>
  <div
    className=""
    style={
      Object {
        "alignItems": "center",
        "background": "#ebebeb",
        "borderRadius": "5px 5px 0 0",
        "display": "flex",
        "gridColumn": "1/4",
        "gridRow": "1/2",
        "justifyContent": "space-between",
        "padding": "20px",
      }
    }
  >
    <h3
      className=""
      style={
        Object {
          "color": "rgba(0,0,0,.7)",
          "fontSize": "1.7em",
          "margin": 0,
          "padding": 0,
        }
      }
    >
      Attention
    </h3>
  </div>
  <div
    currentstep={0}
    style={
      Object {
        "gridColumn": "2/3",
        "gridRow": "2/3",
        "paddingBottom": "20px",
        "paddingTop": "20px",
      }
    }
  />
  <div
    style={
      Object {
        "gridColumn": "2/3",
        "gridRow": "3/4",
        "textAlign": "center",
      }
    }
  />
  <div
    className="mf-footer-area  "
    id="mf-footer"
    style={
      Object {
        "display": "flex",
        "flexDirection": "row",
        "gridColumn": "1/4",
        "gridRow": "4/5",
        "justifyContent": "flex-end",
        "padding": "15px",
      }
    }
  />
</div>
`;

exports[`<Modalfly /> renders an single view modal 1`] = `
<div
  id="mf-wrapper"
  style={
    Object {
      "background": "white",
      "borderRadius": "5px",
      "boxShadow": "0 5px 8px 0 rgba(0, 0, 0, 0.2), 0 7px 20px 0 rgba(0, 0, 0, 0.11)",
      "display": "grid",
      "fontSize": "1em",
      "gridTemplateColumns": "2fr 14fr 2fr",
      "gridTemplateRows": "70px auto 10px 80px",
      "left": "50%",
      "maxHeight": "100%",
      "maxWidth": "100%",
      "minHeight": "35%",
      "opacity": 0,
      "overflow": "visible",
      "position": "fixed",
      "top": "40%",
      "transform": "translate(-50%, -50%)",
      "transition": "opacity 500ms",
      "width": "500px",
    }
  }
>
  <div
    className=""
    style={
      Object {
        "alignItems": "center",
        "background": "#ebebeb",
        "borderRadius": "5px 5px 0 0",
        "display": "flex",
        "gridColumn": "1/4",
        "gridRow": "1/2",
        "justifyContent": "space-between",
        "padding": "20px",
      }
    }
  >
    <h3
      className=""
      style={
        Object {
          "color": "rgba(0,0,0,.7)",
          "fontSize": "1.7em",
          "margin": 0,
          "padding": 0,
        }
      }
    >
      Attention
    </h3>
  </div>
  <div
    currentstep={0}
    style={
      Object {
        "gridColumn": "2/3",
        "gridRow": "2/3",
        "paddingBottom": "20px",
        "paddingTop": "20px",
      }
    }
  >
    <div>
      <h3>
        Would you like to overwrite?
      </h3>
      <p>
        Lorem ipsum dolor sit amet consectetur adipisicing elit.
      </p>
    </div>
  </div>
  <div
    style={
      Object {
        "gridColumn": "2/3",
        "gridRow": "3/4",
        "textAlign": "center",
      }
    }
  />
  <div
    className="mf-footer-area  "
    id="mf-footer"
    style={
      Object {
        "display": "flex",
        "flexDirection": "row",
        "gridColumn": "1/4",
        "gridRow": "4/5",
        "justifyContent": "flex-end",
        "padding": "15px",
      }
    }
  />
</div>
`;
